docs/proposals/cmp-parameters.md

```diff
@@ -0,0 +1,313 @@
+
+#### Use case 2:
+
+As an Argo CD user, I would like to be able to parameterize manifests built by a CMP.
+
+For example, if the Argo CD administrator has installed a CMP which applies a last-mile kustomize overlay to a Helm
+repo, I would like to be able to pass values to the Helm chart without having to manually discover those parameter names
+(in other words, they should show up in the Application UI just like with a native Helm Application). I also shouldn't 
+have to ask my Argo CD admin to modify the CMP to accommodate the values as environment variables.
+
+### Implementation Details/Notes/Constraints
+
+#### Terms
+
+* **Parameter definition**: an instance of a data structure which describes an individual parameter that may be applied
+  to a specific Application. (See the [schema](#parameter-definition-schema) below.)
+* **Parameters announcement**: a list of parameter definitions. (See the [schema](#parameters-announcement-schema) below.)
+
+  "Parameters" is plural because each "announcement" will be a list of multiple parameter definitions.
+* **Parameterized CMP**: a CMP which supports rich parameters (i.e. more than environment variables). A CMP is
+  parameterized if either of these is true:
+  1. its configuration includes the sections consumed by the default CMP server to generate parameters announcements
+  2. it is a fully customized CMP server which implements an endpoint to generate parameters announcements
+
+#### Parameters announcement / parameters serialization format
+
+Parameters announcements should be produced by the CMP as JSON. Use JSON instead of YAML because the tooling is better
+(native JSON libraries, StackOverflow answers about jq, etc.).
+
+Parameters should be set in the manifest as a map of section names to parameter name/value pairs. YAML is used because
+it's easy to read/manipulate in an editor when modifying an Application manifest. We partition by section name so that
+the manifest, to the extent possible, is laid out similarly to the UI.
+
+```yaml
+plugin:
+  parameters:
+    main:
+      - name: values-files
+        value: '["values.yaml"]'
+    Helm Parameters:
+      - name: image
+        values: some.repo:tag
+```
+
+**Note:** I'm not sure whether CRDs allow Map<string, list> types. If not, we should consider flattening the schema to
+a list of objects, each object having a `section` field.
+
+Parameters should be communicated _to_ the CMP as JSON in the same schema as is used in the Application manifest.
+JSON might be a surprising choice considering parameters are represented in the manifest as YAML. But I think JSON makes 
+sense because 1) it's used for parameters announcements (consistency is good) and 2) JSON tooling is better.
+
+#### Parameter definition schema
+
+A parameter definition is an object with following schema:
+
+```go
+type ParameterDefinition struct {
+	// Name is the name of a parameter. (required)
+	Name string `json:"name"`
```

Comment:
Names should be unique, but are they unique only within the section, or should they be globally unique

Reply:
I think they should be unique per-section. If my CMP has a parameter called, say, `address` that's applicable to all applications, I wouldn't want it to conflict if used in conjunction with a Helm chart that has a parameter of the same name.